adapters/operaads/operaads.go

```diff
@@ -0,0 +1,214 @@
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	epTemplate, err := template.New("endpoint").Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &adapter{
+		epTemplate: epTemplate,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	impCount := len(request.Imp)
+	requestData := make([]*adapters.RequestData, 0, impCount)
+	errs := []error{}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	err := checkRequest(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	for _, imp := range request.Imp {
+		requestCopy := *request
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: err.Error(),
+			})
+			continue
+		}
+
+		var operaadsExt openrtb_ext.ImpExtOperaads
+		if err := json.Unmarshal(bidderExt.Bidder, &operaadsExt); err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: err.Error(),
+			})
+			continue
+		}
+
+		err := convertImpression(&imp)
+		if err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: err.Error(),
+			})
+			continue
+		}
+
+		imp.TagID = operaadsExt.PlacementID
+
+		requestCopy.Imp = []openrtb2.Imp{imp}
+		reqJSON, err := json.Marshal(&requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		macro := macros.EndpointTemplateParams{PublisherID: operaadsExt.PublisherID, AccountID: operaadsExt.EndpointID}
+		endpoint, err := macros.ResolveMacros(*a.epTemplate, &macro)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		reqData := &adapters.RequestData{
+			Method:  http.MethodPost,
+			Uri:     endpoint,
+			Body:    reqJSON,
+			Headers: headers,
+		}
+		requestData = append(requestData, reqData)
+	}
+	return requestData, errs
+}
+
+func checkRequest(request *openrtb2.BidRequest) error {
+	if request.Device == nil || len(request.Device.OS) == 0 {
+		return &errortypes.BadInput{
+			Message: "Impression is missing device OS information",
+		}
+	}
+
+	return nil
+}
+
+func convertImpression(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		bannerCopy, err := convertBanner(imp.Banner)
+		if err != nil {
+			return err
+		}
+		imp.Banner = bannerCopy
+	}
+	if imp.Native != nil && imp.Native.Request != "" {
+		v := make(map[string]interface{})
+		err := json.Unmarshal([]byte(imp.Native.Request), &v)
+		if err != nil {
+			return err
+		}
+		_, ok := v["native"]
+		if !ok {
+			body, err := json.Marshal(struct {
+				Native interface{} `json:"native"`
+			}{
+				Native: v,
+			})
+			if err != nil {
+				return err
+			}
+			imp.Native.Request = string(body)
```

Comment:
The race detected by the `validate-merge` job is most likely occurring because `Native` is shared memory. You'll need to make a copy of it before modifying it.